src/test/java/com/google/opensesame/projects/ProjectEntityTest.java

```diff
@@ -0,0 +1,161 @@
+  }
+
+  @Test
+  public void queryByIds() throws IOException {
+    // Expect to receive the ProjectEntities with the corresponding repository IDs.
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    when(request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM))
+        .thenReturn(new String[] {firstMockProject.repositoryId, secondMockProject.repositoryId});
+
+    Collection<ProjectEntity> queryResult = ProjectEntity.queryFromRequest(request, mockResponse);
+    Collection<ProjectEntity> expected = Arrays.asList(firstMockProject, secondMockProject);
+    assertNotNull(queryResult);
+    assertTrue(queryResult.size() == expected.size() && queryResult.containsAll(expected));
+    verify(mockResponse, never()).setStatus(anyInt());
+  }
+
+  @Test
+  public void queryNoIds() throws IOException {
+    // Expect to receive no ProjectEntities because the list of IDs was defined but empty.
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    when(request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM)).thenReturn(new String[] {});
+
+    Collection<ProjectEntity> queryResult = ProjectEntity.queryFromRequest(request, mockResponse);
+    assertNotNull(queryResult);
+    assertTrue(queryResult.size() == 0);
+    verify(mockResponse, never()).setStatus(anyInt());
+  }
+
+  @Test
+  public void queryWithInvalidId() throws IOException {
+    // Expect to receive an error response of 404 (resource not found) because one or more IDs
+    // could not be found within the Datastore.
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    when(request.getParameterValues(ProjectEntity.PROJECT_ID_PARAM))
+        .thenReturn(new String[] {firstMockProject.repositoryId, "invalidID"});
+
+    Collection<ProjectEntity> queryResult = ProjectEntity.queryFromRequest(request, mockResponse);
+    assertNull(queryResult);
+    verify(mockResponse, times(1)).setStatus(HttpServletResponse.SC_NOT_FOUND);
+  }
+
+  @Test
+  public void queryWithoutIds() throws IOException {
+    // Expect to receive all ProjectEntities in order of descending number of mentors because no
+    // list of IDs was supplied.
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+
+    Collection<ProjectEntity> queryResult = ProjectEntity.queryFromRequest(request, mockResponse);
+    ProjectEntity[] expectedQueryResult =
+        new ProjectEntity[] {firstMockProject, secondMockProject, thirdMockProject};
+    assertNotNull(queryResult);
+    assertArrayEquals(expectedQueryResult, queryResult.toArray());
+  }
+
+  @Test
+  public void fromRepositoryId() {
+    // Expect the function to return an existing ProjectEntity from the Datastore because an entity
+    // with the supplied ID already exists.
+
+    ProjectEntity projectEntity =
+        ProjectEntity.fromRepositoryIdOrNew(firstMockProject.repositoryId);
+
+    assertEquals(firstMockProject, projectEntity);
+  }
+
+  @Test
+  public void fromNewRepositoryId() {
```

Comment:
Are you testing the ```Servlet``` and the ```Entity``` in the same file? maybe these should be separate files?

Reply:
I'm testing the helper functions included in the `ProjectEntity.java` class. One is the `ProjectEntity.queryFromRequest` seen above and the other is this function, which is used in the mentors servlet. The intention is not to test the servlet here, but just the unit of functionality in `ProjectEntity` that the servlet relies on.